cmd/intake-receiver/main.go

```diff
@@ -0,0 +1,331 @@
+	agentFileMap *fileMap
+	bufPool      sync.Pool
+	bytesBufPool sync.Pool
+	basePath     string
+	rootResponse string
+}
+
+func (h requestHandler) rootHandler() http.Handler {
+	return logHandler(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/" {
+			rw.WriteHeader(200)
+			rw.Write([]byte(h.rootResponse))
+			return
+		}
+		http.Error(
+			rw,
+			fmt.Sprintf(" %s is not implemented", r.URL.Path),
+			http.StatusNotImplemented,
+		)
+	}))
+}
+
+func (h requestHandler) eventHandler() http.Handler {
+	return logHandler(
+		http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			code, err := h.handleRequest(r)
+			if err != nil {
+				log.Println("failed handling request", code, err.Error())
+				http.Error(rw, err.Error(), code)
+				return
+			}
+			rw.WriteHeader(code)
+		}),
+	)
+}
+
+func (h requestHandler) handleRequest(r *http.Request) (int, error) {
+	buf := h.bufPool.Get().(*bytes.Buffer)
+	defer func() {
+		buf.Reset()
+		h.bufPool.Put(buf)
+	}()
+
+	var err error
+	body := r.Body
+	encoding := r.Header.Get("Content-Encoding")
+	switch encoding {
+	case "deflate":
+		body, err = zlib.NewReader(r.Body)
+	case "gzip":
+		body, err = gzip.NewReader(r.Body)
+	}
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"unable to create compressed reader for %s: %v", encoding, err,
+		)
+	}
+
+	var meta metadata
+	if err := h.processBatch(body, buf, &meta); err != nil {
+		return http.StatusBadRequest, fmt.Errorf("invalid request: %v", err)
+	}
+	if meta.IsEmpty() {
+		return http.StatusBadRequest, errors.New("agent not found in metadata")
+	}
+
+	fileName := filepath.Join(h.basePath, agentFileName(meta))
+	f, err := h.agentFileMap.Get(fileName)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"couldn't retrieve storage file: %s", fileName,
+		)
+	}
+	h.agentFileMap.Set(fileName, f)
+
+	if _, err := io.Copy(f, buf); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf(
+			"failed writing to file: %s: %v", fileName, err,
+		)
+	}
+	return http.StatusAccepted, nil
+}
+
+func (h requestHandler) processBatch(body io.ReadCloser, buf io.Writer, meta *metadata) error {
+	byteBuf := h.bytesBufPool.Get().([]byte)
+	defer func() {
+		byteBuf = byteBuf[:0]
+		h.bufPool.Put(byteBuf)
+	}()
+	defer body.Close()
+	scanner := bufio.NewScanner(body)
+	scanner.Buffer(byteBuf, 0)
+	var decodedMeta bool
+	var err error
+	for scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return err
+		}
+
+		line := scanner.Bytes()
+		// Discard any lines that are at the maximum allowed length and aren't
+		// completely scanned.
+		if len(line) == maxScannerBufSize && line[len(line)-1] != '}' {
+			continue
+		}
+		buf.Write(line)
+		buf.Write([]byte("\n"))
+
+		if decodedMeta {
+			// continue scanning
+			continue
+		}
+		if e := json.Unmarshal(scanner.Bytes(), &meta); e != nil {
+			// TODO(marclop) multierror?
+			err = e
+			// Continue scanning
```

Comment:
is this the behavior in the apm-server, to keep scanning even if one of the lines is malformed?

Reply:
yeah I thought so: https://github.com/elastic/apm-server/blob/4dbe93d223e6098d6f83fc5d1e984570d59cbd39/processor/stream/processor.go#L139-L178